### PR TITLE
[JENKINS-24752] Avoid nodeprovisioner delay after consulting strategies

### PR DESCRIPTION
**Pull request by Vlatombe**
For some cloud implementations (container based), the planned node is almost immediate. With the current implementation, one either has to call https://github.com/jenkinsci/jenkins/blob/bc40ab3ae57178d477fe919b12ed761080df7d6b/core/src/main/java/hudson/slaves/NodeProvisioner.java#L166-L181 but has to wait at least 1 second before doing so. 

The following is changing the behaviour to throttle the calls to `suggestReviewNow` every second instead of ignoring the calls happening within the next second. This allows to process quickly planned nodes without having to wait for a full clock tick.

See the following downstream change as illustration https://github.com/jenkinsci/kubernetes-plugin/pull/515

### CODE REVIEW

File: core/src/main/java/hudson/slaves/NodeProvisioner.java [hunk 1]
```diff
@@ -191,23 +191,21 @@ private void update() {
         provisioningLock.lock();
         try {
             lastSuggestedReview = System.currentTimeMillis();
-
-            // We need to get the lock on Queue for two reasons:
```
Comment (jglick) [inline core/src/main/java/hudson/slaves/NodeProvisioner.java]: Maybe keep this block comment in original indentation to avoid the need for a `-w` diff.

**Review by stephenc — APPROVED** [review]
LGTM but a while since I was involved in this code

**Review by jglick — APPROVED** [review]

**Comment by Vlatombe** [comment]
Noting that this would partly address https://issues.jenkins-ci.org/browse/JENKINS-24752 for the cloud implementations returning instantly, but not the ones returning under a second.

**Comment by Vlatombe** [comment]
I switched to a different aproach which is to throttle the review so that we can honor any call to `suggestReviewNow` within the next second.

**Comment by oleg-nenashev** [comment]
Is it still a draft?

**Comment by Vlatombe** [comment]
@oleg-nenashev This is now ready to review

File: core/src/main/java/hudson/slaves/NodeProvisioner.java [hunk 2]
```diff
@@ -165,19 +167,26 @@ public NodeProvisioner(Label label, LoadStatistics loadStatistics) {
...
      * Give the {@link NodeProvisioner} a hint that now would be a good time to think about provisioning some nodes.
-     * The hint will be ignored if subjected to excessive pestering by callers.
+     * Hints are throttled to one every second.
      *
      * @since 1.415
      */
     public void suggestReviewNow() {
-        if (System.currentTimeMillis() > lastSuggestedReview + TimeUnit.SECONDS.toMillis(1)) {
+        long delay = TimeUnit.SECONDS.toMillis(1) - (System.currentTimeMillis() - lastSuggestedReview);
+        if (delay < 0) {
```
Comment (res0nance) [inline core/src/main/java/hudson/slaves/NodeProvisioner.java]: If you have a queued review and call suggestReviewNow you can possibly call update() twice

**Review by stephenc — COMMENTED** [review]
Is there a risk that if the queue breaks the flag never gets cleared? 

Wouldn’t a better strategy be to have a long that is the expectedNextNanos and schedule if overdue by more than 1s or if more than twice the suggested delay away?

For bonus marks, you could add a hiccup thread to measure current thread scheduling latency and use that average + 2 std dev for deciding if the scheduled execution has vanished?

**Comment by Vlatombe** [comment]
The worst that could happen would be to wait for the next clock tick to reset `queuedReview` to false, in case the scheduled execution has vanished. Is that really worth the trouble?

**Comment by stephenc** [comment]
Maybe the github diff view is not showing where `queuedReview` gets set back to `false`... anyway, was just a thought

**Comment by Vlatombe** [comment]
This one ? https://github.com/jenkinsci/jenkins/pull/4082/files#diff-0558cec054efaecfd6bbee9cbbccdab4R209

**Comment by stephenc** [comment]
Yep on my phone it showed a partial diff for your latest commit and that was omitted

**Review by oleg-nenashev — APPROVED** [review]
Fine with me. Will merge tomorrow if no negative feedback

**Pull request merged**